challenges/cmojidra_Headstorm/cmojidraBackend/data/views.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.shortcuts import render
+from django.http import HttpResponse, JsonResponse, QueryDict
+from django.views.decorators.csrf import csrf_exempt
+
+import json, re
+
+@csrf_exempt
+def index(request):
+
+    # Check if the request type
+    if request.method == 'POST':
+        
+        # Retrieve the data from the json file 
+        raw_data = request.FILES['jsondata'].chunks()
+        data = ''
+        for rd in raw_data:
+            data = rd
+        # Check if the data is formed correctly 
+        try:
+            str_data = list(data.decode('utf-8')[2:-2].split(','))
+        except Exception as e:
+            return JsonResponse({"response":"Malformed Data"})
+
+        flag = 0
+        data = []
+        # Check if the values in the file are numbers
+        for val in str_data:
+            if val.strip().isdigit():                
+                data.append(int(val.strip()))
+            else:
+                flag = 1
+                break
+    
+        if flag == 1:
+            return JsonResponse({"response":"Data format not a list of numbers"})
+
+        # Check if the length of data is 500
+        if len(data) != 500:
+            if len(data) < 500:
+                return JsonResponse({"response":"Length of data less than 500"})
+            return JsonResponse({"response":"Length of Data greater than 500"})
+        
+        # Add data and sort it
+        if 'sorted_data' in request.session:
+            request.session['sorted_data'] = sorted(request.session['sorted_data'].extend(data))
```

Comment:
Can you tell me about your decision to use request session as a data store (vs just keeping the data in memory)?

Reply:
Items stored in the memory as a cache can expire after a certain period, however, items stored in the session are saved until the session ends. 

The items stored in the cache can be removed if the memory gets full and it is not the scalable approach. Moreover, there is an upper limit to the data that can be stored in the memory as the memory is limited.

For load-balancing purposes, the session state can be kept external as in-state servers or SQL servers and it can be shared between different instances of the web app. Likewise, Session data is stored in a database table named django_session. . This thing can not be done for the cache. Moreover, a session usually logs the data for one particular user. Hence, for these reasons, I decided to use the session as a data store instead of keeping the data in memory.

Reply:
So in this instance what is responsible for managing the lifecycle of the session?

e.g. Lets pretend I am a downstream user of this api using curl or something similar. If I send two separate requests does each of them get a new session? If so how would I get my numbers after I from the `get` endpoint after I sent them to the `post` endpoint?

Reply:
In this scenario, the session is not maintained at an individual user level. For simplicity purposes, I stored the information in the Django session and it is maintained by a global variable in the server which will be sent as a response for every server call. Another reason why I used Django session is to ensure that in the future if any other endpoint needs access to this global list (variable) it can be easily read updated by the other endpoints in a consistent manner. Definitely, there are multiple ways to do this but due to the above advantages and experience of using Django variables, I decided to use this approach.

Reply:
Ok so under the hood its a globally shared list - that makes sense.

How would this program handle concurrent requests? Is it thread safe?

Reply:
The session thread is partially thread-safe. Generally, the request will see the session data as a snapshot of it during the first access to the request. On saving, it will overwrite the changes that landed after that moment. So the session state will remain consistent, however, some requests' modifications will be lost.

This property is provided on any framework that uses other storage as a cache mechanism which might be shared between multiple processes that are unlikely to provide the modifications of the objects stored inside without any chance of someone's changes being overwritten,

Reply:
> On saving, it will overwrite the changes that landed after that moment. So the session state will remain consistent, however, some requests' modifications will be lost

Is there anything we can do in our program to ensure that we dont lose changes from requests?

Reply:
To ensure that the data is consistent across all operations, we can use locks to ensure concurrency control.
These locks will always allow simultaneous read access. However, in the case of simultaneous read and write operations, Django provides decorators with which we can add transactions.atomic() to the scope wherever data is being updated. This will ensure that no while the data is being updated no other threads are able to read the partially correct (stale) data.
That being said using this might lead to a slight delay in processing requests if there are a lot of simultaneous read-write requests but this is the tradeoff we'll have to consider in our quest to maintain consistent data throughout the system. Hope this solves your query. Thank you.

Reply:
Yep, it does - thanks!